cpp/src/arrow/compute/kernels/scalar_compare.cc

```diff
@@ -746,6 +787,155 @@ std::shared_ptr<ScalarFunction> MakeScalarMinMax(std::string name,
+    DCHECK_OK(func->AddKernel({in_type, in_type, in_type}, boolean(), std::move(exec)));
+  }
+
+  // Time32 and Time64
+  for (auto unit : {TimeUnit::SECOND, TimeUnit::MILLI}) {
+    InputType in_type(match::Time32TypeUnit(unit));
+    auto exec =
+        GeneratePhysicalInteger<ScalarTernaryEqualTypes, BooleanType, Op>(int32());
+    DCHECK_OK(func->AddKernel({in_type, in_type, in_type}, boolean(), std::move(exec)));
+  }
+  for (auto unit : {TimeUnit::MICRO, TimeUnit::NANO}) {
+    InputType in_type(match::Time64TypeUnit(unit));
+    auto exec =
+        GeneratePhysicalInteger<ScalarTernaryEqualTypes, BooleanType, Op>(int64());
+    DCHECK_OK(func->AddKernel({in_type, in_type, in_type}, boolean(), std::move(exec)));
+  }
+
+  for (const std::shared_ptr<DataType>& ty : BaseBinaryTypes()) {
+    auto exec = GenerateVarBinaryBase<ScalarTernaryEqualTypes, BooleanType, Op>(*ty);
+    DCHECK_OK(func->AddKernel(
+        {InputType::Array(ty), InputType::Scalar(ty), InputType::Scalar(ty)}, boolean(),
+        exec));
+    DCHECK_OK(func->AddKernel(
+        {InputType::Array(ty), InputType::Array(ty), InputType::Array(ty)}, boolean(),
+        std::move(exec)));
+  }
+
+  for (const auto id : {Type::DECIMAL128, Type::DECIMAL256}) {
+    auto exec = GenerateDecimal<ScalarTernaryEqualTypes, BooleanType, Op>(id);
+    DCHECK_OK(func->AddKernel({InputType(id), InputType(id), InputType(id)}, boolean(),
+                              std::move(exec)));
+  }
+
+  {
+    auto exec = ScalarTernaryEqualTypes<BooleanType, FixedSizeBinaryType, Op>::Exec;
+    auto ty = InputType(Type::FIXED_SIZE_BINARY);
+    DCHECK_OK(func->AddKernel({ty, ty, ty}, boolean(), std::move(exec)));
+  }
+
+  return func;
+}
+
+const BetweenOptions* GetDefaultBetweenOptions() {
+  static const auto kBetweenOptions = BetweenOptions::Defaults();
+  return &kBetweenOptions;
+}
+
+const FunctionDoc between_doc{"Check if values are in a range, val betwen a and b",
+                              ("A null on either side emits a null comparison result.\n"
+                               "options are used to specify if the endpoints are\n"
+                               "inclusive, possible values are NEITHER (a<val<b),\n"
+                               "LEFT (a<=val<b), RIGHT (a<val<=b), and the default\n"
+                               "if not specified BOTH (a<=val<=b)"),
+                              {"val", "a", "b"}};
+
+class BetweenMetaFunction : public MetaFunction {
+ public:
+  BetweenMetaFunction()
+      : MetaFunction("between", Arity::Ternary(), &between_doc,
+                     GetDefaultBetweenOptions()) {}
+
+  Result<Datum> ExecuteImpl(const std::vector<Datum>& args,
+                            const FunctionOptions* options,
+                            ExecContext* ctx) const override {
+    const BetweenOptions& between_options = static_cast<const BetweenOptions&>(*options);
+    Datum result;
```

Comment:
I disagree with this approach. It defeats the purpose of having the `BetweenOptions` because they are being consumed by `CallFunction`.

I will take a stab at supporting internal dispatching based on `BetweenOptions` and submit a PR to your branch. Up to now this is looking good but needs a bit of tweaking.

Reply:
Thanks for your feedback. Ok, on moving the functions to within the dispatch for between.  Underlying APIs should support natural integration into other programming languages and not force Pandas/Python approach, but also be easy to maintain and at least consistent for related functions. Would it make sense to then also use one compare function, but have functions on top of this for `greater`, `less` etc.? The `not_between` function could then be implemented by calling `between` and reversing the result.

I understand that Pandas has lot of influence. In addition to databases, there are other libraries such as
- [Ruby Daru](https://github.com/ankane/rover)
- [R Dplyr](https://dplyr.tidyverse.org/reference/between.html)
- [Hive Between](https://sqlandhadoop.com/hive-between/)
- [Ruby Rover](https://github.com/ankane/rover)
- [Ruby CSV table](https://ruby-doc.org/stdlib-3.1.0/libdoc/csv/rdoc/CSV/Table.html)
- [Link compare](https://docs.microsoft.com/en-us/dotnet/visual-basic/language-reference/operators/comparison-operators)
- [Apache Calcite Algebra](https://calcite.apache.org/docs/algebra.html)

For many of these, chained operations is the way to implement between. It is unclear if one can write a Kernel generator that would minimize loading data from memory for chained operations - Calcite approach seems useful, but need to understand it better. Will ask on mailing list for thoughts on these kind of approaches since they would be something new.

Finally, I have not done much performance measurement and optimization of these kernels, but this can probably be done once tests and relevant functionality are working.